refactor:  PXE - Replace old NoteDataProvider tests with new

wildjos opened the pull request:
## Description:
The existing `note_data_provider_tests.ts` file relied heavily on parameterised arrays (e.g., filteringTests) and top-level helpers, which made it difficult to quickly understand which scenarios were covered. 

This PR replaces the old file with a complete rewrite.
**Heads up:** Git shows this as a modification because of the file path, but treat it as a brand-new test. 


## What’s changed
- Deleted the old test file.  
- Added new test file
- Test helpers (`mkNote`, `mkNullifier`, `setupProviderWithNotes`)
- Covers the same functional areas as before, with added coverage for edge cases:  
  - Creation of `NoteDataProvider`  
  - `getNotes` filtering (contract, slot, scope, status)  
  - `applyNullifiers` happy path and edge cases  
  - Rollback of notes and nullifiers  
- Ensured resources (db) closed after use

## Bug fixes to make tests work with current implementation
1. `getNotes` now throws an `Error` if passed an empty `scopes` array.  
2. Filtering nullified notes by scope fixed: previously, `getNotes(..., scopes: [...])` returned all nullified notes regardless of scope; now it correctly respects the scope filter.  

## Testing summary
- All previous scenarios retained and verified in new tests.  
- Happy-path and edge-case coverage for filtering, nullification, and rollback fully implemented.  
- Confirms behaviour is correct across multiple scopes, multiple contracts, storage slots, and nullification states.

See attached test logs for before and after 
[old_test.log](https://github.com/user-attachments/files/22849038/old_test.log)
[new_test.log](https://github.com/user-attachments/files/22849039/new_test.log)

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
 import { Fr } from '@aztec/foundation/fields';
-import { openTmpStore } from '@aztec/kv-store/lmdb-v2';
+// import { openTmpStore } from '@aztec/kv-store/lmdb-v2';
+// import { LmdbStore } from '@aztec/kv-store/lmdb-v2';
```
```suggestion
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -251,6 +251,11 @@ export class NoteDataProvider {
 
+    // throw early if scopes is an empty array
+    if (filter.scopes !== undefined && filter.scopes.length === 0) {
+      throw new Error('Trying to get notes with an empty scopes array');
```
```suggestion
      throw new Error('Trying to get notes with an empty scopes array. Scopes have to be set to undefined if intending on not filtering by scopes.');
```
From the original error message it would not be clear how exactly to fix the error.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```suggestion
   * @throws Error if trying to access notes for a scope not in the PXE database
   * @throws If filtering by an empty scopes array. Scopes have to be set to undefined or to a non-empty array.
```

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+  describe('NoteDataProvider.getNotes filtering happy path', () => {
+    let store: AztecLMDBStoreV2;
+    let provider: NoteDataProvider;
+    let note1: NoteDao; // eslint-disable-line @typescript-eslint/no-unused-vars
```
```suggestion
```

Would just drop all the unused vars in the test. Declaring them is pointless.

wildjos replied:
TLDR: the tests don't work if I remove "unused vars"

The setup function: `setupProviderWithNotes` returns note1, note2 and note3. 
If I remove `let note1` from the test scope, TypeScript complains because I'm destructuring into something that doesn't exist.  
If i leave it in, the linter complains that the variable isn't used in assertions. 
But they still need to be there to recieve values during setup.  And those values are needed for the test logic to work correctly.

So in this case the eslint-disable flags have come to the rescue - I will add a comment to make it clear.

benesjan replied:
When destructuring an object you don't have to use all the object fields and doing this change compiles:

<img width="654" height="248" alt="image" src="https://github.com/user-attachments/assets/e204d112-511d-4975-aeb5-6d632dda6a09" />

wildjos replied:
🤯. Will update!

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+      scope2 = AztecAddress.fromString('0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb');
+      slotX = Fr.fromString('0x01');
+      slotY = Fr.fromString('0x02');
+    });
```
Given that these are always set to the same values would just drop this block and have thees values as constants.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+
+      const noteA = await mkNote({ contractAddress: contractA, recipient: scope1, index: BigInt(1) });
+      const noteB = await mkNote({ contractAddress: contractB, recipient: scope2, index: BigInt(2) });
+      await provider1.addNotes([noteA, noteB], AztecAddress.ZERO);
```
```suggestion
      const scope = await AztecAddress.random();
      await provider1.addNotes([noteA, noteB], scope);
```
Just a detail but I think it's better to use random value here as zero address made me wonder whether zero scope has some special meaning.

wildjos replied:
I have a weird paranoia about random() even though the chances a collision or failure is astronomically small.   I am adding a constant to make it clear that this is a valid (non-zero) address, and the test will remain deterministic.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+
+    it('filters notes matching only the contractAddress', async () => {
+      const res = await provider.getNotes({ contractAddress: contractA });
+      expect(res.map(n => n.index.toString()).sort()).toEqual(['1', '2']);
```
Would declare a constant for the the note indexes and use it both in the setup function and here as it's not super clear where these values came from.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+
+    it('filters notes matching only the contractAddress', async () => {
+      const res = await provider.getNotes({ contractAddress: contractA });
+      expect(res.map(n => n.index.toString()).sort()).toEqual(['1', '2']);
```
```suggestion
      expect(res.map(n => n.index).sort()).toEqual([1n, 2n]);
```

The conversion to `string` is unnecessary here.

wildjos replied:
Thanks - so I removed the `toString()` conversion. Sorting didn’t behave as I expected, which broke my tests because they were comparing strings '1', '2' rather than the BigNums 1n, 2n....  Which made me reconsider what the expect's are actually checking. 

- The DB returns notes in arbitrary order.
- Sorting BigInts without a comparator gave this:
```
const indexes = [1n, 10n, 2n, 4n];
indexes.sort(); 
// expected: [1n, 2n, 4n, 10n]
// got:      [1n, 10n, 2n, 4n]
```

I figured it’s better not to rewrite the expected results.   And adding a comparator for BigInt would just add more chaff.

So I updated the tests to use `Set` for equality checks and `.toContain` / `.not.toContain` where appropriate. This also ensures we’re checking _exactly_ the contents, not just “does it contain this… might also contain more.”

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+    });
+
+    it('returns no notes when filtering by non-existing contractAddress', async () => {
+      const res = await provider.getNotes({ contractAddress: AztecAddress.ZERO });
```
Would also use random address here as that makes it clear there is no special meaning and that we are not expected to get a match.

wildjos replied:
Using FAKE_ADDRESS

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+    });
+
+    it('returns no notes when filtering by non-existing storageSlot', async () => {
+      const res = await provider.getNotes({ contractAddress: contractA, storageSlot: Fr.fromString('0x00') });
```
Same for the storage slot here - would use random value.

wildjos replied:
Changed to NON_EXISTING_SLOT

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+      expect(activeNotes.map(n => n.index.toString()).sort()).toEqual(['1', '2']);
+    });
+
+    it('updates all relevant indexes when nullifying notes', async () => {
```
This test title is confusing as it makes me think of note index. It also doesn't seem to check all the relevant indexes/maps as it doesn't check `#nullifiedNotesByTxHash` and `#nullifiedNotesByRecipient`.

wildjos replied:
Good - spot on the missing indexes/maps. 

However, my follow-one PR will remove `txHash` and `byRecipient`.  Therefore I have not included specific tests.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+
+      // should throw on second attempt as note1 is already nullified
+      await expect(provider.applyNullifiers([mkNullifier(note1)])).rejects.toThrow(
+        'Nullifier not found in applyNullifiers',
```
In the `applyNullifiers` function I would check if the note is already nullified (we can just look if the note exists in `#nullifiedNotesByNullifier`) and if yes would throw a nicer error for it.

wildjos replied:
Good idea, adding this.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+
+    describe('rewind nullifications happy path', () => {
+      async function setupRollbackScenario() {
+        const noteBlock1 = await mkNote({ index: BigInt(1), l2BlockNumber: 1 }); // Nullified at block 2
```
```suggestion
        const noteBlock1 = await mkNote({ index: 1n, l2BlockNumber: 1 }); // Nullified at block 2
```
Just a detail but we generally use this syntax for instantiating a bigint in the repo and I think it's easier to read.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+        expect(allNotes.map(n => n.index.toString())).toEqual(['10']);
+      });
+
+      it('restores note when nullified after the rollback block', async () => {
```
This doesn't seem like an edge case and it's already tested by the "restores notes that were nullified after the rollback block" test so I would just drop this.

wildjos replied:
ok

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
-      await expect(
-        noteDataProvider.applyNullifiers([
+    describe('rewind nullifications edge cases', () => {
+      it('handles rollback when blockNumber equals synchedBlockNumber', async () => {
```
Unrelated to this PR but I think it would be a good idea to just drop the `synchedBlockNumber` from the function arguments of `rollbackNotesAndNullifiers` in note data provider. in that function we provide it because we don't have the info on the block upper bound in which there could be a nullifier and what makes it worse is that we have that info for notes.

I think we should store that info for nullifiers as well, drop the arg and then we can drop these 3 edge cases tests.

This seems like a good example of a complexity leak.

Would you agree?

wildjos replied:
Agreed — this needs reworking.

I’ll address it in a follow-on PR after the filtering/index updates, along with the related pruning/bounding discussion with @nventuro  and @Thunkar .

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+        expect(activeNotes.map(n => n.index.toString())).toEqual(['11']); // Note should be restored
+      });
+
+      it('handles rollback when synchedBlockNumber < blockNumber', async () => {
```
I think this might indicate malfunction in `Synchronizer` but would not worry about that now because as mentioned above would just drop the problematic synchedBlockNumber arg.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts:
```diff
@@ -1,469 +1,581 @@
+        expect(allNotes.map(n => n.index.toString())).toEqual(['20']);
+      });
+
+      it('handles rollback with a large block gap', async () => {
```
Also this test could be dropped if we drop the `synchedBlockNumber` arg.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```suggestion
        throw new Error('Trying to get incoming notes of a scope that is not in the PXE database');
```
Unrelated but it was bothering me and we are already touching the file in this PR so it seems fine to sneak the fix in here.

wildjos replied:
Sorry - lost the trail... which fix to apply here?

benesjan replied:
"an scope" -> "a scope"

wildjos replied:
Got it!  Thanks :)

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -293,17 +298,32 @@ export class NoteDataProvider {
-              : await toArray(this.#nullifiedNotesByContract.getValuesAsync(filter.contractAddress.toString())),
-        notes: this.#nullifiedNotes,
-      });
+    // Nullified note sources (filtered by scope)
```
```suggestion
    // If status is ACTIVE_OR_NULLIFIED we add nullified notes as candidates on top of the default active ones.
```

benesjan reviewed (approved):
Looks good!

Feel free to merge once you address my feedback.

Thanks

wildjos committed:
refactor:  PXE - Replace old NoteDataProvider tests with new

## Description:
The existing `note_data_provider_tests.ts` file relied heavily on parameterised arrays (e.g., filteringTests) and top-level helpers, which made it difficult to quickly understand which scenarios were covered.

This PR replaces the old file with a complete rewrite.
**Heads up:** Git shows this as a modification because of the file path, but treat it as a brand-new test.

## What’s changed
- Deleted the old test file.
- Added new test file
- Test helpers (`mkNote`, `mkNullifier`, `setupProviderWithNotes`)
- Covers the same functional areas as before, with added coverage for edge cases:
  - Creation of `NoteDataProvider`
  - `getNotes` filtering (contract, slot, scope, status)
  - `applyNullifiers` happy path and edge cases
  - Rollback of notes and nullifiers
- Ensured resources (db) closed after use

## Bug fixes to make tests work with current implementation
1. `getNotes` now throws an `Error` if passed an empty `scopes` array.
2. Filtering nullified notes by scope fixed: previously, `getNotes(..., scopes: [...])` returned all nullified notes regardless of scope; now it correctly respects the scope filter.

## Testing summary
- All previous scenarios retained and verified in new tests.
- Happy-path and edge-case coverage for filtering, nullification, and rollback fully implemented.
- Confirms behaviour is correct across multiple scopes, multiple contracts, storage slots, and nullification states.

See attached test logs for before and after
[old_test.log](https://github.com/user-attachments/files/22849038/old_test.log)
[new_test.log](https://github.com/user-attachments/files/22849039/new_test.log)
